### PR TITLE
fix buggy logic in type resolution of type suffixes. #34

### PR DESCRIPTION
**Pull request by SSoulaimane**
issue #34

### CODE REVIEW

File: src/dsymbol/tests.d [hunk 1]
```diff
@@ -66,3 +69,77 @@ unittest
...
+    return getTokensForParser(source.dup.representation, config, &stringCache);
+}
+
+unittest
+{
+    import std.meta : AliasSeq;
+
+    auto tokens = lex(q{int a = 9;});
+    foreach(i, t;
+        AliasSeq!(tok!"int", tok!"identifier", tok!"=", tok!"intLiteral", tok!";"))
```
Comment (BBasile) [inline src/dsymbol/tests.d]: You can use an array literal here, there is no need for the values to be available at compile time

File: src/dsymbol/tests.d [hunk 2]
```diff
@@ -66,3 +69,77 @@ unittest
...
+    }
+    assert(tokens[1].text == "a", tokens[1].text);
+    assert(tokens[3].text == "9", tokens[3].text);
+}
+
+string randomDFilename()
+{
+    import std.uuid : randomUUID;
+    import std.string : translate;
+    return "dsymbol_" ~ randomUUID().toString().translate(['-': '_']) ~ ".d";
```
Comment (BBasile) [inline src/dsymbol/tests.d]: you don't need to translate.

File: src/dsymbol/tests.d [hunk 3]
```diff
@@ -15,7 +18,7 @@ import std.stdio : writeln, stdout;
  *      results = An array of string array. Each slot represents the variable name
  *      followed by the type strings.
  */
-version (unittest)
+version (unittest):
 void expectSymbolsAndTypes(const string source, const string[][] results)
 {
     static immutable rName = "dsymbol-test-154251542-56564105-78944416-98523170-02452336.d";
```
Comment (BBasile) [inline src/dsymbol/tests.d]: why not replacing this with a `randomDFilename()` ?

File: src/dsymbol/conversion/second.d [hunk 4]
```diff
@@ -298,6 +301,53 @@ body
 	}
 }
 
+unittest
```
Comment (BBasile) [inline src/dsymbol/conversion/second.d]: Why don't you use the existing testing system ? It's more compact and does the same.

File: src/dsymbol/tests.d (hunk 1 above)
Comment (BBasile) [inline src/dsymbol/tests.d]: Dismiss this, since this comment i suggest to change the tests in another PR

File: src/dsymbol/tests.d (hunk 2 above)
Comment (BBasile) [inline src/dsymbol/tests.d]: Dismiss this, since this comment i suggest to change the tests in another PR

File: src/dsymbol/tests.d (hunk 3 above)
Comment (BBasile) [inline src/dsymbol/tests.d]: Dismiss this, since this comment i suggest to change the tests in another PR

**Comment by BBasile** [comment]
#36 passes, let's find why. The only difference is that i use the existing tester routine.

**Comment by BBasile** [comment]
#37 created a conflict and forces you to rebase and remove your tests. I've even added the tests for POINTER_SYMBOL_NAME, which just have to be un-commented.

**Comment by BBasile** [comment]
There's a serious problem with this PR actually, it breaks completion of pointers in DCD. For now DCD follows D implicit dereference. You can see the effect with   ```int* a.|```. Before the change we got completion for an int, after we got nothing.

The right behavior would be to ignore the first POINTER_SYMBOL_NAME since for example even the compiler allows this  `int* a; writeln(a.max)`. If you add an indirection it doesn't work anymore. Currently D symbol always gives the completions, whatever is the indirections count.

File: src/dsymbol/conversion/second.d (hunk 4 above)
Comment (SSoulaimane) [inline src/dsymbol/conversion/second.d]: True. I like the declarative approach, it is more compact.
but the alternative is more flexible, more things are testable for potential bugs, for instance: symbol.kind, symbol.qualifier , symbol.isPointer ...etc.
Maybe we should see if we can make it the current system more flexible.

**Comment by SSoulaimane** [comment]
> #36 passes, let's find why. The only difference is that i use the existing tester routine.
that is one weird bug that apparently only affects dmd >= 2.075.
i got the solution/workaround though.

File: src/dsymbol/tests.d (hunk 1 above)
Comment (SSoulaimane) [inline src/dsymbol/tests.d]: sure

**Comment by BBasile** [comment]
You didn't address the issue that is the that first indirection should not cause DCD not to give the completions. Try your modified file in DCD and you'll see that 

```d
struct Foo{int a;}
Foo* foo;
foo. // dot completion here
```

does not give results anymore.

**Comment by SSoulaimane** [comment]
> You didn't address the issue that is the that first indirection should not cause DCD not to give the completions.

I'm definitely going to fix that in dcd. its should be trivial. I've been doing a lot of changes in my local dcd fork, among these some refactoring of symbol swapping with symbol.type which is the real cause of this bug, also added swapping eponymous templates with their content.

the reason I need pointers to be distinguished is because I've been progressing with type checking in dcd for ufcs but the code relies on many changes to dsymbol which I would like to push first.

I there's a few bugs I've found which some I've fixed locally and some not, I'll report them one by one and push the fixes with some tests.

**Comment by BBasile** [comment]
Looks like it's in a dead end, closing.

**Pull request closed without merging**